lib/cache/OperationWriter.js

```diff
@@ -0,0 +1,255 @@
+    this.stopped = null;
+    this._resolveStopped = null;
+    // tracks operation hashes for duplicate prevention
+    this.recentOperationHashes = new LRU({max: 1000, maxAge: 5 * 60 * 1000});
+    // operation queue collection
+    this.collection = database.collections.continuity2017_operation_queue;
+    // queue length in database
+    this.databaseOperationQueueLength = 0;
+    const operationsConfig = config['ledger-consensus-continuity'].operations;
+    // max operation queue length
+    this.MAX_OPERATION_QUEUE_LENGTH = operationsConfig.maxQueueSize;
+  }
+
+  /**
+   * Starts this OperationWriter. Any operations that have been added to its
+   * queue will be bulk inserted when the queue fills or when a timeout occurs.
+   */
+  start() {
+    const {queue, running, highWaterMark, timeout} = this;
+    if(running) {
+      // already running
+      return;
+    }
+
+    // now running
+    this.running = true;
+
+    // if the queue is at or over the highWaterMark, schedule immediate flush
+    if(queue.length >= highWaterMark) {
+      this._scheduleFlush({timeout: 0});
+    } else if(queue.length > 0) {
+      // queue is not empty, schedule a timeout flush
+      this._scheduleFlush({timeout});
+    }
+
+    this.stopped = new Promise(res => {
+      this._resolveStopped = res;
+    });
+  }
+
+  /**
+   * Stops this OperationWriter. Events can still be added to the queue, but
+   * they will not be automatically bulk inserted.
+   */
+  stop() {
+    const {running, timeoutId} = this;
+    if(!running) {
+      // already not running
+      return;
+    }
+    // cancel any scheduled flush
+    if(timeoutId !== null) {
+      clearTimeout(timeoutId);
+      this.timeoutId = null;
+    }
+    this.running = false;
+    this.stopped = null;
+    this._resolveStopped();
+  }
+
+  /**
+   * Adds an operation to be written. If this OperationWriter is running and the
+   * operation causes the queue to reach the configured `highWaterMark`, then
+   * it will automatically flush operations to the underlying database via a
+   * bulk insert.
+   *
+   * @param {object} options - Options to use.
+   * @param {object} options.operation - The operation to write.
+   * @param {object} options.meta - The metadata to write.
+   */
+  async add({operation, meta, forceFlush = false} = {}) {
+    const {queue, running, highWaterMark, MAX_OPERATION_QUEUE_LENGTH} = this;
+
+    if(this.databaseOperationQueueLength >= MAX_OPERATION_QUEUE_LENGTH) {
+      // we intentionally do not await the promise, the length can be updated
+      // after the error is returned
+      this._cacheDatabaseQueueLength().catch(e => {
+        logger.debug(DATABASE_QUEUE_LENGTH_ERROR, {error: e});
+      });
+      throw new BedrockError(
+        'The node is not accepting operations. Try again later.',
+        'OperationError', {
+          httpStatusCode: 503,
+          public: true,
+        });
+    }
+
+    if(this._isDuplicateOperation(meta)) {
```

Comment:
It would be good to mention that this check is a soft duplicate check and that other processes could have accepted the same operation -- and, if appropriate, mention that the same process could accept the same operation if another one was pushed out of the LRU cache. Perhaps the explanation will be that the max queue length needs to be less than the LRU cache size for this to work properly -- but I'll note that the max queue length seems to be a soft check as well, so there seems like there could be bugs at the edges.